### PR TITLE
Monitor memory usage

### PR DESCRIPTION
**Pull request by bitdivine**
# Motivation
We would like to have visibility into memory usage.

# Changes
* Add stable and wasm memory to the exported metrics.
  * Note: We do not make much use of stable memory yet, except in upgrades, so this will currently show the stable memory populated by the last `pre_upgrade` hook.  However we plan to use stable memory more in future.
  * Note: Given that memory size thresholds are in GiB, I have provided this data as MiB so that it is easy to read at a glance without calculation.  The stats API cannot do the same as the stats do not support f64, as it is does not support Eq.
* Add memory logging to the `pre_upgrade` hook, so we have the very latest data available if an upgrade goes wrong.
* Add wasm as the target in for `cargo build`, `cargo check` (covered by `.cargo/config -> build`) and `cargo clippy`, as the wasm memory size code compiles only when wasm is the target.

# Tests
* A test that all expected metrics are present is included in this PR.  It does not test that the values are plausible or e.g. how much memory increases with each user.
* The `get_stats` API call is exercised in `estimate-upgrade-cycles`.  This PR also re-enables the stats calls made during the downgrade-upgrade test.  No use is made of the data though, so this just confirms that the API call works, not that the data is correct.
* This branch has been deployed to an app subnet.  The metrics show the heap size: https://xnjld-hqaaa-aaaal-qb56q-cai.raw.icp0.io/metrics

# Todos

- [x] Add entry to changelog (if necessary).

# Future
- Estimates of memory usage if a given wasm is deployed to production.

### CODE REVIEW

File: rs/backend/src/stats.rs [hunk 1]
```diff
@@ -33,6 +42,11 @@ pub struct Stats {
     pub neurons_topped_up_count: u64,
     pub transactions_to_process_queue_length: u32,
     pub performance_counts: Vec<PerformanceCount>,
+    // TODO: After a transition period, these two can be required rather than being optional.
+    //       The transition period can be considered over when most deployments, including
+    //       production, CI and snsdemo populate these fields.
+    pub stable_memory_size_bytes: Option<u64>,
```
Comment (lmuntaner) [inline rs/backend/src/stats.rs:48]: Is this stored in stable memory in pre-upgrade?

Is this why you need the optional and keep backwards compatibility?

**Review by lmuntaner — APPROVED** [review]
LGTM! Just one comment.

File: rs/backend/src/stats.rs (hunk 1 above)
Comment (bitdivine) [reply rs/backend/src/stats.rs:48]: These are not stored in stable memory.  Yes, it is for backwards compatibility but the backwards compatibility is to allow tests to run against new and old versions of the code.  Previously, in such cases, some tests were simply disabled for a while.  I don't believe that that is a great solution so I'm exploring what can be done without disabling tests.

Comment (bitdivine) [reply rs/backend/src/stats.rs:48]: Happy to discuss further but closing so that I can merge.

Comment (lmuntaner) [reply rs/backend/src/stats.rs:48]: Ok, sounds good! Thanks for the explanation

**Pull request merged**